metaseq/tasks/streaming_CM3_language_modeling.py

```diff
@@ -0,0 +1,232 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates. All Rights Reserved.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+"""
+Streaming Language Modeling task that loads corpora in plaintext and performs
+on-the-fly tokenization.
+"""
+
+import logging
+import os
+from dataclasses import dataclass, field
+from typing import Any, Dict, List, Optional
+
+import numpy as np
+import torch
+from omegaconf import II
+
+from metaseq.data import (
+    Dictionary,
+    CausalMaskedDataset,
+    PartitionedStreamingDataset,
+    iterators
+)
+from metaseq.tasks.streaming_language_modeling import StreamingLanguageModelingConfig, StreamingLanguageModelingTask
+from metaseq.tasks import register_task
+
+try:
+    from tokenizers import ByteLevelBPETokenizer
+    has_hf_tokenizers = True
+except ImportError:
+    has_hf_tokenizers = False
+
+
+logger = logging.getLogger(__name__)
+
+IMAGE_PREFIX = "I"
+SPEECH_PREFIX = "S"
+
+
+@dataclass
+class StreamingCM3LanguageModelingConfig(StreamingLanguageModelingConfig):
+    image_tokens: int = field(
+        default=1024,
+        metadata={"help": "total number of vision tokens used"},
+    )
+    speech_tokens: int = field(
+        default=1024,
+        metadata={"help": "total number of speech tokens used"},
+    )
+    num_sentinel_tokens: int = field(
+        default=512,
+        metadata={"help": "number of special sentinel tokens to add to the vocabulary"},
+    )
+    lambda_sentinel_tokens: int = field(
+        default=1,
+        metadata={"help": "poisson lambda for the cm3 objective"},
+    )
+
+
+@register_task("streaming_CM3_language_modeling", dataclass=StreamingCM3LanguageModelingConfig)
+class StreamingCM3LanguageModelingTask(StreamingLanguageModelingTask):
+    def __init__(self, args):
+        self.args = args
+        self.datasets = {}
+        self.dataset_to_epoch_iter = {}
+
+        if not has_hf_tokenizers:
+            raise ImportError("Please install tokenizers with: pip install tokenizers")
+
+        tokenizer = ByteLevelBPETokenizer.from_file(
+            args.vocab_filename,
+            args.merges_filename
+        )
+        self.sentinel_tokens = [f"<sentinel:{i}>" for i in range(args.num_sentinel_tokens)]
+        self.sentinel_end = "<eoss>"
+
+        tokenizer.add_special_tokens([f"{IMAGE_PREFIX}{x} " for x in range(args.image_tokens)])
+        tokenizer.add_special_tokens([f"{SPEECH_PREFIX}{x} " for x in range(args.speech_tokens)])
+        tokenizer.add_special_tokens(self.sentinel_tokens)
+        tokenizer.add_special_tokens([self.sentinel_end])
```

Comment:
HF has a way for us to move all of this into one tokenizer file that contains the full pipeline. Should we do that?

Reply:
(I thought we had chosen wordpiece? I'm indifferent but)